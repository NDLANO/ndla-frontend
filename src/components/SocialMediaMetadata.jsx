/**
 * Copyright (c) 2019-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React from 'react';
import Helmet from 'react-helmet';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import { BasenameContext } from '../App';
import config from '../config';
import {
  LocationShape,
  ImageShape,
  ArticleShape,
  LearningpathShape,
} from '../shapes';
import { getHtmlLang, appLocales } from '../i18n';

export const getAlternateLanguages = (basename, locale, trackableContent) => {
  const defaultLocale = getHtmlLang();
  const isBasenamePage = locale === defaultLocale && basename === '';
  if (!trackableContent && isBasenamePage) {
    return appLocales.map(appLocale => appLocale.abbreviation);
  }
  if (
    (!trackableContent && !isBasenamePage) ||
    !isBasenamePage ||
    !trackableContent.supportedLanguages ||
    trackableContent.supportedLanguages.length === 0
  ) {
    return [];
  }
  return trackableContent.supportedLanguages.filter(
    language =>
      !!appLocales.find(appLocale => appLocale.abbreviation === language),
  );
};

export const SocialMediaMetadata = ({
  title,
  image,
  description,
  locale,
  trackableContent,
  location,
  children,
}) => (
  <BasenameContext.Consumer>
    {basename => (
      <Helmet>
        <link
          rel="canonical"
          href={`${config.ndlaFrontendDomain}${location.pathname}`}
        />
        {getAlternateLanguages(basename, locale, trackableContent).map(
          alternateLanguage => (
            <link
              key={alternateLanguage}
              rel="alternate"
              hrefLang={alternateLanguage}
              href={`${config.ndlaFrontendDomain}/${alternateLanguage}${location.pathname}`}
            />
          ),
        )}
        {children}
        <meta property="og:type" content="article" />
        <meta name="twitter:card" content="summary" />
        <meta name="twitter:site" content="@ndla_no" />
        <meta name="twitter:creator" content="@ndla_no" />
        <meta
          property="og:url"
          content={`${config.ndlaFrontendDomain}${location.pathname}`}
        />
        {title && <meta property="og:title" content={`${title} - NDLA`} />}
        {title && <meta name="twitter:title" content={`${title} - NDLA`} />}
        {description && (
          <meta property="og:description" content={description} />
        )}
        {description && (
          <meta name="twitter:description" content={description} />
        )}
        {image && image.src && <meta property="og:image" content={image.src} />}
        {image && image.src && (
          <meta name="twitter:image:src" content={image.src} />
        )}
        {!image || !image.src ? (
          <meta
            name="twitter:image:src"
            content={`${config.ndlaFrontendDomain}/static/metalogo.jpg`}
          />
        ) : (
          ''
        )}
        {!image || !image.src ? (
          <meta
            property="og:image"
            content={`${config.ndlaFrontendDomain}/static/metalogo.jpg`}
          />
        ) : (
          ''
        )}
        <meta property="og:site_name" content="ndla.no" />
        <meta
          property="article:publisher"
          content="https://www.facebook.com/ndla.no"
        />
        <meta
          property="article:author"
          content="https://www.facebook.com/ndla.no"
        />
      </Helmet>
    )}
  </BasenameContext.Consumer>
);

SocialMediaMetadata.propTypes = {
  title: PropTypes.string,
  description: PropTypes.string,
  locale: PropTypes.string,
  location: LocationShape,
  image: ImageShape,
  trackableContent: PropTypes.oneOfType([ArticleShape, LearningpathShape]),
};

export default withRouter(SocialMediaMetadata);
