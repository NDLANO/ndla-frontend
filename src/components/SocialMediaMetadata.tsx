/**
 * Copyright (c) 2019-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { ReactNode } from 'react';
import { Helmet } from 'react-helmet-async';
import { useLocation, Location, useParams } from 'react-router-dom';
import config from '../config';
import { preferredLocales, isValidLocale } from '../i18n';

export const getCanonicalUrl = (location: Location) => {
  if (!location.pathname.includes('article-iframe')) {
    return `${config.ndlaFrontendDomain}${location.pathname}`;
  }
  const paths = location.pathname.split('/');
  if (isValidLocale(paths[2])) {
    paths.splice(2, 1);
  }
  return `${config.ndlaFrontendDomain}${paths.join('/')}`;
};

export const getAlternateUrl = (
  location: Location,
  alternateLanguage: string,
) => {
  if (!location.pathname.includes('article-iframe')) {
    return `${config.ndlaFrontendDomain}/${alternateLanguage}${location.pathname}`;
  }
  const paths = location.pathname.split('/');
  if (isValidLocale(paths[2])) {
    paths.splice(2, 1);
  }
  paths.splice(2, 0, alternateLanguage);
  return `${config.ndlaFrontendDomain}${paths.join('/')}`;
};

export const getAlternateLanguages = (trackableContent?: TrackableContent) => {
  if (!trackableContent || !trackableContent.supportedLanguages) {
    return preferredLocales.map((appLocale) => appLocale.abbreviation);
  }
  if (trackableContent?.supportedLanguages?.length === 0) {
    return [];
  }
  return trackableContent.supportedLanguages.filter((language) =>
    isValidLocale(language),
  );
};

export const getOgUrl = (location: Location, basename?: string) => {
  const ogBaseName = !basename?.length ? '' : `/${basename}`;
  return `${config.ndlaFrontendDomain}${ogBaseName}${location.pathname}`;
};

interface TrackableContent {
  tags?: string[];
  supportedLanguages?: string[];
}

interface Props {
  title: string;
  description?: string;
  imageUrl?: string;
  audioUrl?: string;
  trackableContent?: TrackableContent;
  children?: ReactNode;
  type?: string;
}

const SocialMediaMetadata = ({
  title,
  imageUrl,
  audioUrl,
  description,
  trackableContent,
  children,
  type = 'article',
}: Props) => {
  const location = useLocation();
  const { lang } = useParams();
  return (
    <Helmet>
      <link rel="canonical" href={getCanonicalUrl(location)} />
      {getAlternateLanguages(trackableContent).map((alternateLanguage) => (
        <link
          key={alternateLanguage}
          rel="alternate"
          hrefLang={alternateLanguage}
          href={getAlternateUrl(location, alternateLanguage)}
        />
      ))}
      {children}
      {trackableContent?.tags && (
        <meta property="keywords" content={`${trackableContent?.tags}`} />
      )}
      <meta property="og:type" content={type} />
      <meta name="twitter:card" content="summary" />
      <meta name="twitter:site" content="@ndla_no" />
      <meta name="twitter:creator" content="@ndla_no" />
      <meta property="og:url" content={getOgUrl(location, lang)} />
      {title && <meta property="og:title" content={`${title} - NDLA`} />}
      {title && <meta name="twitter:title" content={`${title} - NDLA`} />}
      {description && <meta property="og:description" content={description} />}
      {description && <meta name="description" content={description} />}
      {description && <meta name="twitter:description" content={description} />}
      {audioUrl && <meta property="og:audio" content={audioUrl} />}
      {imageUrl && <meta property="og:image" content={imageUrl} />}
      {imageUrl && <meta name="twitter:image:src" content={imageUrl} />}
      {!imageUrl ? (
        <meta
          name="twitter:image:src"
          content={`${config.ndlaFrontendDomain}/static/metalogo.jpg`}
        />
      ) : (
        ''
      )}
      {!imageUrl ? (
        <meta
          property="og:image"
          content={`${config.ndlaFrontendDomain}/static/metalogo.jpg`}
        />
      ) : (
        ''
      )}
      <meta property="og:site_name" content="ndla.no" />
    </Helmet>
  );
};

export default SocialMediaMetadata;
