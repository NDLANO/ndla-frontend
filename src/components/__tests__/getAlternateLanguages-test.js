/*
 * Copyright (c) 2019-present, NDLA.
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { getAlternateLanguages } from '../SocialMediaMetadata';

test('getAlternateLanguages with article and basename is empty', () => {
  const alternateLanguages = getAlternateLanguages('', 'nb', {
    supportedLanguages: ['nb', 'nn', 'en'],
  });
  expect(alternateLanguages).toMatchSnapshot();
});

test('getAlternateLanguages with article and basename is en', () => {
  const alternateLanguages = getAlternateLanguages('en', 'en', {
    supportedLanguages: ['nb', 'nn', 'en'],
  });
  expect(alternateLanguages).toMatchSnapshot();
});

test('getAlternateLanguages without article and basename is empty', () => {
  const alternateLanguages = getAlternateLanguages('', 'nb');
  expect(alternateLanguages).toMatchSnapshot();
});

test('getAlternateLanguages without article and basename is nn', () => {
  const alternateLanguages = getAlternateLanguages('nn', 'nn');
  expect(alternateLanguages).toMatchSnapshot();
});
