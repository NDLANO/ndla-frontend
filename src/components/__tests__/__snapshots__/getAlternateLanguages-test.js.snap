// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAlternateLanguages with article and basename is empty 1`] = `
Array [
  "nb",
  "nn",
  "en",
]
`;

exports[`getAlternateLanguages with article and basename is en 1`] = `Array []`;

exports[`getAlternateLanguages without article and basename is empty 1`] = `
Array [
  "nb",
  "nn",
  "en",
]
`;

exports[`getAlternateLanguages without article and basename is nn 1`] = `Array []`;
